lxd/firewall/drivers/drivers_xtables.go

```diff
@@ -0,0 +1,518 @@
+		if err != nil {
+			return err
+		}
+
+	} else {
+		err := d.iptablesPrepend(family, comment, "nat", "POSTROUTING", args...)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// NetworkSetupDHCPDNSAccess sets up basic iptables overrides for DHCP/DNS.
+func (d XTables) NetworkSetupDHCPDNSAccess(networkName string, ipVersion uint) error {
+	var rules [][]string
+	if ipVersion == 4 {
+		rules = [][]string{
+			{"4", networkName, "filter", "INPUT", "-i", networkName, "-p", "udp", "--dport", "67", "-j", "ACCEPT"},
+			{"4", networkName, "filter", "INPUT", "-i", networkName, "-p", "udp", "--dport", "53", "-j", "ACCEPT"},
+			{"4", networkName, "filter", "INPUT", "-i", networkName, "-p", "tcp", "--dport", "53", "-j", "ACCEPT"},
+			{"4", networkName, "filter", "OUTPUT", "-o", networkName, "-p", "udp", "--sport", "67", "-j", "ACCEPT"},
+			{"4", networkName, "filter", "OUTPUT", "-o", networkName, "-p", "udp", "--sport", "53", "-j", "ACCEPT"},
+			{"4", networkName, "filter", "OUTPUT", "-o", networkName, "-p", "tcp", "--sport", "53", "-j", "ACCEPT"}}
+	} else if ipVersion == 6 {
+		rules = [][]string{
+			{"6", networkName, "filter", "INPUT", "-i", networkName, "-p", "udp", "--dport", "547", "-j", "ACCEPT"},
+			{"6", networkName, "filter", "INPUT", "-i", networkName, "-p", "udp", "--dport", "53", "-j", "ACCEPT"},
+			{"6", networkName, "filter", "INPUT", "-i", networkName, "-p", "tcp", "--dport", "53", "-j", "ACCEPT"},
+			{"6", networkName, "filter", "OUTPUT", "-o", networkName, "-p", "udp", "--sport", "547", "-j", "ACCEPT"},
+			{"6", networkName, "filter", "OUTPUT", "-o", networkName, "-p", "udp", "--sport", "53", "-j", "ACCEPT"},
+			{"6", networkName, "filter", "OUTPUT", "-o", networkName, "-p", "tcp", "--sport", "53", "-j", "ACCEPT"}}
+	} else {
+		return fmt.Errorf("Invalid IP version")
+	}
+
+	comment := d.networkIPTablesComment(networkName)
+
+	for _, rule := range rules {
+		ipVersion, err := strconv.ParseUint(rule[0], 10, 0)
+		if err != nil {
+			return err
+		}
+
+		err = d.iptablesPrepend(uint(ipVersion), comment, rule[2], rule[3], rule[4:]...)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// NetworkSetupDHCPv4Checksum attempts a workaround for broken DHCP clients.
+func (d XTables) NetworkSetupDHCPv4Checksum(networkName string) error {
+	comment := d.networkIPTablesComment(networkName)
+	return d.iptablesPrepend(4, comment, "mangle", "POSTROUTING", "-o", networkName, "-p", "udp", "--dport", "68", "-j", "CHECKSUM", "--checksum-fill")
+}
+
+// NetworkClear removes network rules.
+func (d XTables) NetworkClear(networkName string, ipVersion uint) error {
+	err := d.iptablesClear(ipVersion, d.networkIPTablesComment(networkName), "mangle")
+	if err != nil {
+		return err
+	}
+
+	err = d.iptablesClear(ipVersion, d.networkIPTablesComment(networkName), "filter")
+	if err != nil {
+		return err
+	}
+
+	err = d.iptablesClear(ipVersion, d.networkIPTablesComment(networkName), "nat")
+	if err != nil {
+		return err
+	}
```

Comment:
Right so as mentioned in our chat this morning, this can be a problem if some of those tables are missing.
That's why in the past we had to line up the logic adding the rules to the logic removing it, so we effectively fail during adding time instead of potentially failing due to a missing table on clearing time.

We should either re-introduce the old logic or try to check whether the tables exist.
Something like this may work:
```stgraber@castiana:~/data/code/lxc/lxd (stgraber/storage)$ sudo cat /proc/self/net/ip_tables_names
nat
mangle
filter
stgraber@castiana:~/data/code/lxc/lxd (stgraber/storage)$ sudo cat /proc/self/net/ip6_tables_names
nat
filter
```

I don't know how recent or reliable that mechanism is, but we could at least attempt to get the list that way and if we fail, assume they're all present.